### ISSUE TITLE
How are the quantiles used in training the models

### ISSUE THREAD

**Issue by fengcong1992** [cbuelt/forecasting-challenge#1]
Hi there,

Thanks for organizing this code/data repository so well. I have some questions regarding the quantiles. It seems you  used quantiles in training the models (calculating pinball loss and get results). But the model output during the training is always the same deterministic (quantile 50?), which means even though the code has the quantile output option (e.g., in the train_wind_temp_model() function), but it doesn't take a quantile in the input (i.e., train_data). Please let me know if I understand the code in a wrong way. And please kindly let me know how did you calculate quantiles to train the models. Thanks.

**Comment by cbuelt** [comment]
Hello, thanks for your comment. I hope I understood everything right.
Let's take the case of the wind_temp_model(), which I used for wind and temperature data.
The model outputs all five quantiles (0.025,0.25,0.5,0.75,0.975) by using a Dense layer of size 5 as the last layer. The pinball loss then evaluates all five quantiles simultaneously.

So you are right, the model does not take the quantile as an input. However it is able to learn to predict the true quantile. This is because the pinball loss function is optimized by the respective quantile (Much like the mean squared error is minimized by the mean). So by optimizing the neural network using this specific loss function, it learns to forecast a quantile (in timestep t+1) in dependence of the data available (in timestep t).
I hope this explanation can help you, otherwise feel free to reach out again.

**Comment by fengcong1992** [comment]
> Hello, thanks for your comment. I hope I understood everything right. Let's take the case of the wind_temp_model(), which I used for wind and temperature data. The model outputs all five quantiles (0.025,0.25,0.5,0.75,0.975) by using a Dense layer of size 5 as the last layer. The pinball loss then evaluates all five quantiles simultaneously.
> 
> So you are right, the model does not take the quantile as an input. However it is able to learn to predict the true quantile. This is because the pinball loss function is optimized by the respective quantile (Much like the mean squared error is minimized by the mean). So by optimizing the neural network using this specific loss function, it learns to forecast a quantile (in timestep t+1) in dependence of the data available (in timestep t). I hope this explanation can help you, otherwise feel free to reach out again.

Dear Chris,

Thanks for the prompt reply. And my apologize for the confusing question. It totally makes sense now. But I am still confused about how you get the quantiles (as the output) for training at the training stage. By running the wind_forecasting.ipynb code, training_target shape is (n,1), which means the output is a single time series, not the 5 quantiles. Did I misunderstood your data preprocessing/formatting? My question is how did you get the optimal quantiles from your data (single time series, or ensemble members?) at the training stage for training? Thanks!

**Comment by cbuelt** [comment]
Yes, you are absolutely right, the training_target shape is (n,1), while the model output shape is (n,5). However the pinball loss (as implemented in tensorflow) is able to directly evaluate the loss for all five different quantiles as a vector.
This is basically the same as taking the mean over five different quantile losses of shape (n,5). 
This work was created in a university course. Our aim was to forecast the five mentioned quantiles. You can view the forecasts as as five separate time series, only with the addition that they were all created by the same model. Each single time series (n,1) is then compared and evaluated with the training_target of shape (n,1). So in a way, the model predicts five different time series at the same time, each consisting of forecasts for different quantiles.

Also I am not quite sure if I understand you correctly, regarding the optimal quantiles. The optimal quantiles are unknown and not used in the process. 
To be fair, the course in which I developed this model was some time ago, so I'm sorry if I don't perfectly get the points you're referring to.
Hope that still helps.
Chris

**Comment by fengcong1992** [comment]
> Yes, you are absolutely right, the training_target shape is (n,1), while the model output shape is (n,5). However the pinball loss (as implemented in tensorflow) is able to directly evaluate the loss for all five different quantiles as a vector. This is basically the same as taking the mean over five different quantile losses of shape (n,5). This work was created in a university course. Our aim was to forecast the five mentioned quantiles. You can view the forecasts as as five separate time series, only with the addition that they were all created by the same model. Each single time series (n,1) is then compared and evaluated with the training_target of shape (n,1). So in a way, the model predicts five different time series at the same time, each consisting of forecasts for different quantiles.
> 
> Also I am not quite sure if I understand you correctly, regarding the optimal quantiles. The optimal quantiles are unknown and not used in the process. To be fair, the course in which I developed this model was some time ago, so I'm sorry if I don't perfectly get the points you're referring to. Hope that still helps. Chris

Yes, I am very familiar with the probabilistic forecasting evaluation metrics and how to use deep learning predict multiple quantiles in one model. But to train a model with five quantiles, your training_target should be (n,5) to be consistent with the the model output (Dense(5)), right? My question was how did you get the 5 quantiles from your data for training? I feel I miss an important link in the workflow. Thanks.

**Comment by cbuelt** [comment]
I get what you mean. In principle, the training target should be consistent (n,5). However, it would be five times the same value, as we are predicting the same timestep, only different quantiles. Now one solution would be to just extend the training target to five times the same value, which would be not very memory efficient.
Instead, the pinball loss function automatically evaluates the five quantiles against the one output. So the function basically gets a list of quantiles (n_q), input of shape (n,1) and predictions of shape (n,n_q). Now for each quantile the predictions are shape (n,1) and are compared against the target with shape (n,1). The different values are than averaged over all quantiles.

So instead of having a redundant training target of (n,5), we can have a shape of (n,1). It works the same, but is more memory efficient.

**Comment by fengcong1992** [comment]
> I get what you mean. In principle, the training target should be consistent (n,5). However, it would be five times the same value, as we are predicting the same timestep, only different quantiles. Now one solution would be to just extend the training target to five times the same value, which would be not very memory efficient. Instead, the pinball loss function automatically evaluates the five quantiles against the one output. So the function basically gets a list of quantiles (n_q), input of shape (n,1) and predictions of shape (n,n_q). Now for each quantile the predictions are shape (n,1) and are compared against the target with shape (n,1). The different values are than averaged over all quantiles.
> 
> So instead of having a redundant training target of (n,5), we can have a shape of (n,1). It works the same, but is more memory efficient.

Dear Chris,

After a few days' thinking and testing, I think you are right. Thanks for clarifying the implementation. And great job on maintaining this repository.
